Lib/fontTools/subset/svg.py

```diff
@@ -0,0 +1,235 @@
+    # and local url(#...) links found in fill or clip-path attributes
+    # TODO(anthrotype): Check we aren't missing other supported kinds of reference
+    find_svg_elements_with_references = etree.XPath(
+        ".//svg:*[ "
+        "starts-with(@xlink:href, '#') "
+        "or starts-with(@fill, 'url(#') "
+        "or starts-with(@clip-path, 'url(#') "
+        "or contains(@style, ':url(#') "
+        "]",
+        namespaces=NAMESPACES,
+    )
+
+    find_svg_elements_with_glyph_href = etree.XPath(
+        ".//svg:*[starts-with(@xlink:href, '#glyph')]", namespaces=NAMESPACES
+    )
+
+
+def group_elements_by_id(tree: etree.Element) -> Dict[str, etree.Element]:
+    return {el.attrib["id"]: el for el in find_svg_elements_with_id(tree)}
+
+
+def parse_css_declarations(style_attr: str) -> Dict[str, str]:
+    # https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/style
+    # https://developer.mozilla.org/en-US/docs/Web/CSS/Syntax#css_declarations
+    result = {}
+    for declaration in style_attr.split(";"):
+        if declaration.count(":") == 1:
+            property_name, value = declaration.split(":")
+            property_name = property_name.strip()
+            result[property_name] = value.strip()
+        elif declaration.strip():
+            raise ValueError(f"Invalid CSS declaration syntax: {declaration}")
+    return result
+
+
+def iter_referenced_ids(tree: etree.Element) -> Iterator[str]:
+    # Yield all the ids that can be reached via references from within this element tree
+    for el in chain([tree], find_svg_elements_with_references(tree)):
+        if XLINK_HREF in el.attrib:
+            href = el.attrib[XLINK_HREF]
+            if href.startswith("#"):
+                ref_id = href[1:]
+                assert ref_id
+                yield ref_id
+
+        attrs = el.attrib
+        if "style" in attrs:
+            attrs = {**attrs, **parse_css_declarations(el.attrib["style"])}
+        for attr in ("fill", "clip-path"):
+            if attr in attrs:
+                value = attrs[attr]
+                if value.startswith("url(#") and value.endswith(")"):
+                    ref_id = value[5:-1]
+                    assert ref_id
+                    yield ref_id
+
+
+def closure_element_ids(
+    elements: Dict[str, etree.Element], element_ids: Set[str]
+) -> None:
+    # Expand the initial subset of element ids to include ids that can be reached
+    # via references from the initial set.
+    unvisited = element_ids
+    while unvisited:
+        referenced: Set[str] = set()
+        for el_id in unvisited:
+            if el_id not in elements:
+                # ignore dangling reference; not our job to validate svg
+                continue
+            referenced.update(iter_referenced_ids(elements[el_id]))
+        referenced -= element_ids
+        element_ids.update(referenced)
+        unvisited = referenced
+
+
+def subset_elements(el: etree.Element, ids: Set[str]) -> bool:
```

Comment:
IIUC ids might be clearer as something like retained_ids

Reply:
Done 92470ea